File: server/services/generate-contract-pdf.js
/**
 * Genererar en professionell kontrakts-PDF för RankLeague influencer-avtal.
 * Använder PDFKit (ren Node.js — ingen Python behövs).
 */

import PDFDocument from 'pdfkit';
import fs from 'fs';

// Färger
const PURPLE = '#7c3aed';
const DARK_PURPLE = '#5b21b6';
const LIGHT_PURPLE = '#ede9fe';
const DARK_TEXT = '#111827';
const GRAY_TEXT = '#6b7280';
const LIGHT_GRAY = '#f3f4f6';
const BORDER_GRAY = '#d1d5db';
const WHITE = '#ffffff';

export function generateContractPdf(data, outputPath) {
  return new Promise((resolve, reject) => {
    const doc = new PDFDocument({
      size: 'A4',
      margins: { top: 60, bottom: 60, left: 70, right: 70 },
    });

    const stream = fs.createWriteStream(outputPath);
    doc.pipe(stream);

    const pageWidth = doc.page.width - doc.page.margins.left - doc.page.margins.right;

    // === HEADER ===
    doc.fontSize(10).fillColor(PURPLE).font('Helvetica-Bold')
      .text('RANKLEAGUE', { align: 'center' });

    doc.moveDown(0.3);
    doc.fontSize(24).fillColor(DARK_TEXT).font('Helvetica-Bold')
      .text('SAMARBETSAVTAL', { align: 'center' });

    doc.moveDown(0.3);
    doc.fontSize(11).fillColor(GRAY_TEXT).font('Helvetica')
      .text(`Mellan ${data.foretag_namn} och ${data.influencer_namn}`, { align: 'center' });

    doc.moveDown(0.5);

    // Lila linje
    const lineY = doc.y;
    doc.moveTo(doc.page.margins.left, lineY)
      .lineTo(doc.page.width - doc.page.margins.right, lineY)
      .strokeColor(PURPLE).lineWidth(2).stroke();

    doc.moveDown(1);

    // === 1. AVTALSPARTER ===
    sectionHeader(doc, '1. Avtalsparter');

    // Tabell-header
    const tableX = doc.page.margins.left;
    const col1 = 90;
    const col2 = (pageWidth - col1) / 2;
    const col3 = (pageWidth - col1) / 2;
    const rowH = 24;

    let ty = doc.y;

    // Header-rad
    drawRow(doc, ty, tableX, [col1, col2, col3], ['', 'Uppdragsgivare', 'Influencer'], {
      bg: PURPLE, textColor: WHITE, bold: true
    });
    ty += rowH;

    const parterRows = [
      ['Namn', data.foretag_namn || '', data.influencer_namn || ''],
      ['Kontaktperson', data.kontaktperson || '', ''],
      ['E-post', data.foretag_epost || '', data.influencer_epost || ''],
      ['Kanal', '', `@${data.kanalnamn || ''} (${data.plattform || ''})`],
      ['Referral-kod', '', data.referral_kod || ''],
    ];

    for (let i = 0; i < parterRows.length; i++) {
      drawRow(doc, ty, tableX, [col1, col2, col3], parterRows[i], {
        bg: i % 2 === 0 ? LIGHT_GRAY : WHITE,
        textColor: DARK_TEXT,
        labelColor: DARK_PURPLE,
      });
      ty += rowH;
    }

    doc.y = ty + 12;

    // === 2. UPPDRAGSBESKRIVNING ===
    sectionHeader(doc, '2. Uppdragsbeskrivning');
    bodyText(doc, pageWidth,
      `Influencern ska producera och publicera videoinnehåll på ${data.plattform || 'sin kanal'} ` +
      `där ${data.foretag_namn}s tjänst RankLeague (rankleague.com) presenteras ` +
      `och marknadsförs mot influencerns publik.`
    );

    // === 3. ERSÄTTNING ===
    sectionHeader(doc, '3. Ersättning');

    ty = doc.y;
    const eCol1 = 170;
    const eCol2 = 120;
    const eCol3 = pageWidth - eCol1 - eCol2;

    drawRow(doc, ty, tableX, [eCol1, eCol2, eCol3], ['Typ', 'Belopp', 'Detaljer'], {
      bg: DARK_PURPLE, textColor: WHITE, bold: true
    });
    ty += rowH;

    const perVideo = data.per_video_sek || 300;
    const perSignup = data.per_signup_sek || 10;
    const maxVideos = data.videos_required || 5;

    const ersRows = [
      ['Fast ersättning per video', `${perVideo} SEK`, `Max ${maxVideos} videos`],
      ['Provision per signup', `${perSignup} SEK`, `Via referral-kod: ${data.referral_kod || ''}`],
      ['Max fast ersättning', `${perVideo * maxVideos} SEK`, `${maxVideos} videos x ${perVideo} SEK`],
    ];

    for (let i = 0; i < ersRows.length; i++) {
      drawRow(doc, ty, tableX, [eCol1, eCol2, eCol3], ersRows[i], {
        bg: i % 2 === 0 ? LIGHT_PURPLE : WHITE,
        textColor: DARK_TEXT,
      });
      ty += rowH;
    }

    doc.y = ty + 12;

    // === 4. KRAV ===
    sectionHeader(doc, '4. Krav på innehåll');

    const krav = [
      `Varje video MÅSTE innehålla en tydlig och stark call-to-action (CTA) som uppmanar tittarna att registrera sig på RankLeague via influencerns referral-länk eller referral-kod.`,
      `CTA:n ska vara verbal (sägs högt i videon) och visuell (visas på skärmen).`,
      `Influencern ska namnge RankLeague och kort förklara tjänstens värde.`,
      `Referral-koden ${data.referral_kod || ''} eller referral-länken ska vara tydligt synlig i videon och i videobeskrivningen.`,
      `Innehållet får inte strida mot svensk lag eller RankLeagues värderingar.`,
    ];
    for (const item of krav) {
      bulletItem(doc, pageWidth, item);
    }

    // === 5. RAPPORTERING ===
    checkPageSpace(doc, 150);
    sectionHeader(doc, '5. Rapportering');
    bodyText(doc, pageWidth,
      'Influencern ska inom 7 dagar efter varje publicerad video dela följande statistik med Uppdragsgivaren:'
    );
    bulletItem(doc, pageWidth, 'Antal visningar');
    bulletItem(doc, pageWidth, 'Antal klick på referral-länk (om tillgängligt)');
    bulletItem(doc, pageWidth, 'Skärmbild på videostatistik från plattformens analys');

    // === 6. AVTALSTID ===
    checkPageSpace(doc, 100);
    sectionHeader(doc, '6. Avtalstid');
    const avtalstid = data.avtalstid_dagar || 30;
    bodyText(doc, pageWidth,
      `Avtalet gäller i ${avtalstid} dagar från och med signeringsdatum. ` +
      `Om inget annat överenskommits upphör avtalet automatiskt efter avtalstiden.`
    );
    if (data.expires_at) {
      bodyText(doc, pageWidth, `Utgångsdatum: ${data.expires_at.slice(0, 10)}`);
    }

    // === 7. BETALNINGSVILLKOR ===
    checkPageSpace(doc, 100);
    sectionHeader(doc, '7. Betalningsvillkor');
    bodyText(doc, pageWidth,
      'Utbetalning sker inom 30 dagar efter att influencern levererat statistik för respektive video. ' +
      'Provisionsersättning (signups) utbetalas månadsvis baserat på data från RankLeagues system.'
    );

    // === SIGNERING ===
    checkPageSpace(doc, 200);
    doc.moveDown(1);

    // Grå linje
    const sigLineY = doc.y;
    doc.moveTo(doc.page.margins.left, sigLineY)
      .lineTo(doc.page.width - doc.page.margins.right, sigLineY)
      .strokeColor(BORDER_GRAY).lineWidth(0.5).stroke();

    doc.moveDown(0.8);
    sectionHeader(doc, 'Signering');
    bodyText(doc, pageWidth, 'Genom att signera detta avtal godkänner båda parter ovanstående villkor.');

    const datum = data.datum || new Date().toISOString().split('T')[0];
    bodyText(doc, pageWidth, `Datum: ${datum}`);

    doc.moveDown(1.5);

    // Signerings-kolumner
    const sigX1 = doc.page.margins.left;
    const sigX2 = doc.page.margins.left + pageWidth / 2 + 20;
    const sigY = doc.y;

    doc.fontSize(10).fillColor(DARK_TEXT).font('Helvetica');
    doc.text('Uppdragsgivare:', sigX1, sigY);
    doc.text(data.foretag_namn || '', sigX1, sigY + 14);
    doc.text(data.kontaktperson || '', sigX1, sigY + 28);

    doc.text('Influencer:', sigX2, sigY);
    doc.text(data.influencer_namn || '', sigX2, sigY + 14);
    doc.text(`@${data.kanalnamn || ''}`, sigX2, sigY + 28);

    doc.moveDown(4);
    const underY = sigY + 70;

    doc.text('____________________________', sigX1, underY);
    doc.text('____________________________', sigX2, underY);

    doc.fontSize(8).fillColor(GRAY_TEXT);
    doc.text('Underskrift', sigX1, underY + 14);
    doc.text('Underskrift', sigX2, underY + 14);

    // Footer
    doc.moveDown(3);
    const footY = doc.y;
    doc.moveTo(doc.page.margins.left, footY)
      .lineTo(doc.page.width - doc.page.margins.right, footY)
      .strokeColor(BORDER_GRAY).lineWidth(0.5).stroke();

    doc.moveDown(0.5);
    // SparkCollab-branding (diskret, ljusgrå, 8pt)
    doc.fontSize(8).fillColor(GRAY_TEXT).font('Helvetica')
      .text(`Genererat via SparkCollab.se  •  ${datum}`, { align: 'center' });

    // Klart
    doc.end();

    stream.on('finish', () => resolve(outputPath));
    stream.on('error', reject);
  });
}


// === Hjälpfunktioner ===

function sectionHeader(doc, text) {
  doc.moveDown(0.6);
  doc.fontSize(13).fillColor(DARK_PURPLE).font('Helvetica-Bold')
    .text(text);
  doc.moveDown(0.3);
}

function bodyText(doc, width, text) {
  doc.fontSize(10).fillColor(DARK_TEXT).font('Helvetica')
    .text(text, { width, lineGap: 3 });
  doc.moveDown(0.3);
}

function bulletItem(doc, width, text) {
  const x = doc.x;
  doc.fontSize(10).fillColor(DARK_TEXT).font('Helvetica')
    .text(`•  ${text}`, doc.page.margins.left + 16, doc.y, { width: width - 16, lineGap: 2 });
  doc.moveDown(0.15);
}

function checkPageSpace(doc, needed) {
  if (doc.y + needed > doc.page.height - doc.page.margins.bottom) {
    doc.addPage();
  }
}

function drawRow(doc, y, x, colWidths, cells, opts = {}) {
  const rowH = 24;

  // Bakgrund
  if (opts.bg) {
    doc.rect(x, y, colWidths.reduce((a, b) => a + b, 0), rowH)
      .fill(opts.bg);
  }

  // Ram
  doc.rect(x, y, colWidths.reduce((a, b) => a + b, 0), rowH)
    .strokeColor(BORDER_GRAY).lineWidth(0.5).stroke();

  // Text i varje cell
  let cx = x;
  for (let i = 0; i < cells.length; i++) {
    const isLabel = i === 0 && opts.labelColor;
    doc.fontSize(9)
      .fillColor(isLabel ? opts.labelColor : (opts.textColor || DARK_TEXT))
      .font(opts.bold || isLabel ? 'Helvetica-Bold' : 'Helvetica')
      .text(cells[i] || '', cx + 8, y + 7, {
        width: colWidths[i] - 16,
        height: rowH - 4,
        ellipsis: true,
      });
    cx += colWidths[i];
  }
}

File: server/services/generate-contract-pdf.py
#!/usr/bin/env python3
"""
Genererar en professionell kontrakts-PDF för RankLeague influencer-avtal.
Anropas med JSON via stdin, skriver PDF till angiven output-sökväg.

Användning:
  echo '{"output_path": "kontrakt.pdf", ...}' | python3 generate-contract-pdf.py
"""

import json
import sys
from datetime import datetime

from reportlab.lib.pagesizes import A4
from reportlab.lib.units import mm, cm
from reportlab.lib.colors import HexColor
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.lib.enums import TA_LEFT, TA_CENTER, TA_RIGHT
from reportlab.platypus import (
    SimpleDocTemplate, Paragraph, Spacer, Table, TableStyle,
    HRFlowable, KeepTogether
)


# Färger
PURPLE = HexColor('#7c3aed')
DARK_PURPLE = HexColor('#5b21b6')
LIGHT_PURPLE = HexColor('#ede9fe')
DARK_BG = HexColor('#1f2937')
GRAY_TEXT = HexColor('#6b7280')
DARK_TEXT = HexColor('#111827')
WHITE = HexColor('#ffffff')
LIGHT_GRAY = HexColor('#f3f4f6')
BORDER_GRAY = HexColor('#d1d5db')


def build_styles():
    styles = getSampleStyleSheet()

    styles.add(ParagraphStyle(
        'ContractTitle',
        parent=styles['Title'],
        fontSize=22,
        textColor=DARK_TEXT,
        spaceAfter=6,
        alignment=TA_CENTER,
        fontName='Helvetica-Bold',
    ))

    styles.add(ParagraphStyle(
        'ContractSubtitle',
        parent=styles['Normal'],
        fontSize=11,
        textColor=GRAY_TEXT,
        spaceAfter=20,
        alignment=TA_CENTER,
        fontName='Helvetica',
    ))

    styles.add(ParagraphStyle(
        'SectionHeader',
        parent=styles['Heading2'],
        fontSize=13,
        textColor=DARK_PURPLE,
        spaceBefore=16,
        spaceAfter=8,
        fontName='Helvetica-Bold',
        borderPadding=(0, 0, 4, 0),
    ))

    styles.add(ParagraphStyle(
        'BodyText2',
        parent=styles['Normal'],
        fontSize=10,
        textColor=DARK_TEXT,
        spaceAfter=6,
        fontName='Helvetica',
        leading=14,
    ))

    styles.add(ParagraphStyle(
        'BulletItem',
        parent=styles['Normal'],
        fontSize=10,
        textColor=DARK_TEXT,
        spaceAfter=4,
        fontName='Helvetica',
        leftIndent=16,
        leading=14,
    ))

    styles.add(ParagraphStyle(
        'SmallGray',
        parent=styles['Normal'],
        fontSize=8,
        textColor=GRAY_TEXT,
        fontName='Helvetica',
    ))

    styles.add(ParagraphStyle(
        'SignatureLine',
        parent=styles['Normal'],
        fontSize=10,
        textColor=DARK_TEXT,
        fontName='Helvetica',
        spaceBefore=30,
        spaceAfter=4,
    ))

    return styles


def generate_pdf(data):
    output_path = data['output_path']
    styles = build_styles()

    doc = SimpleDocTemplate(
        output_path,
        pagesize=A4,
        topMargin=2 * cm,
        bottomMargin=2 * cm,
        leftMargin=2.5 * cm,
        rightMargin=2.5 * cm,
    )

    story = []

    # === HEADER ===
    story.append(Paragraph("RANKLEAGUE", ParagraphStyle(
        'Brand', parent=styles['Normal'],
        fontSize=10, textColor=PURPLE, fontName='Helvetica-Bold',
        alignment=TA_CENTER, spaceAfter=4,
    )))

    story.append(Paragraph("SAMARBETSAVTAL", styles['ContractTitle']))

    subtitle = f"Mellan {data.get('foretag_namn', 'Foretag')} och {data.get('influencer_namn', 'Influencer')}"
    story.append(Paragraph(subtitle, styles['ContractSubtitle']))

    # Linje
    story.append(HRFlowable(
        width="100%", thickness=2, color=PURPLE,
        spaceBefore=0, spaceAfter=16
    ))

    # === AVTALSPARTER (tabell) ===
    story.append(Paragraph("1. Avtalsparter", styles['SectionHeader']))

    parter_data = [
        ['', 'Uppdragsgivare', 'Influencer'],
        ['Namn', data.get('foretag_namn', ''), data.get('influencer_namn', '')],
        ['Kontaktperson', data.get('kontaktperson', ''), ''],
        ['E-post', data.get('foretag_epost', ''), data.get('influencer_epost', '')],
        ['Kanal', '', f"@{data.get('kanalnamn', '')} ({data.get('plattform', '')})"],
        ['Referral-kod', '', data.get('referral_kod', '')],
    ]

    parter_table = Table(parter_data, colWidths=[80, 190, 190])
    parter_table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (-1, 0), PURPLE),
        ('TEXTCOLOR', (0, 0), (-1, 0), WHITE),
        ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
        ('FONTSIZE', (0, 0), (-1, -1), 9),
        ('FONTNAME', (0, 1), (0, -1), 'Helvetica-Bold'),
        ('TEXTCOLOR', (0, 1), (0, -1), DARK_PURPLE),
        ('BACKGROUND', (0, 1), (-1, -1), LIGHT_GRAY),
        ('ALIGN', (0, 0), (-1, -1), 'LEFT'),
        ('VALIGN', (0, 0), (-1, -1), 'MIDDLE'),
        ('GRID', (0, 0), (-1, -1), 0.5, BORDER_GRAY),
        ('TOPPADDING', (0, 0), (-1, -1), 6),
        ('BOTTOMPADDING', (0, 0), (-1, -1), 6),
        ('LEFTPADDING', (0, 0), (-1, -1), 8),
    ]))
    story.append(parter_table)
    story.append(Spacer(1, 12))

    # === UPPDRAGSBESKRIVNING ===
    story.append(Paragraph("2. Uppdragsbeskrivning", styles['SectionHeader']))
    story.append(Paragraph(
        f"Influencern ska producera och publicera videoinnehall pa {data.get('plattform', 'sin kanal')} "
        f"dar {data.get('foretag_namn', 'Uppdragsgivaren')}s tjanst RankLeague (rankleague.com) presenteras "
        f"och marknadsf\u00f6rs mot influencerns publik.",
        styles['BodyText2']
    ))

    # === ERSÄTTNING ===
    story.append(Paragraph("3. Ersattning", styles['SectionHeader']))

    ersattning_data = [
        ['Typ', 'Belopp', 'Detaljer'],
        ['Fast ersattning per video', f"{data.get('per_video_sek', 300)} SEK", f"Max {data.get('videos_required', 5)} videos"],
        ['Provision per signup', f"{data.get('per_signup_sek', 10)} SEK", f"Via referral-kod: {data.get('referral_kod', '')}"],
        ['Max fast ersattning', f"{data.get('per_video_sek', 300) * data.get('videos_required', 5)} SEK", f"{data.get('videos_required', 5)} videos x {data.get('per_video_sek', 300)} SEK"],
    ]

    ers_table = Table(ersattning_data, colWidths=[160, 120, 180])
    ers_table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (-1, 0), DARK_PURPLE),
        ('TEXTCOLOR', (0, 0), (-1, 0), WHITE),
        ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
        ('FONTSIZE', (0, 0), (-1, -1), 9),
        ('BACKGROUND', (0, 1), (-1, -1), LIGHT_PURPLE),
        ('ALIGN', (1, 1), (1, -1), 'CENTER'),
        ('GRID', (0, 0), (-1, -1), 0.5, BORDER_GRAY),
        ('TOPPADDING', (0, 0), (-1, -1), 6),
        ('BOTTOMPADDING', (0, 0), (-1, -1), 6),
        ('LEFTPADDING', (0, 0), (-1, -1), 8),
    ]))
    story.append(ers_table)
    story.append(Spacer(1, 12))

    # === KRAV ===
    story.append(Paragraph("4. Krav pa innehall", styles['SectionHeader']))
    krav = [
        "Varje video <b>MASTE</b> innehalla en tydlig och stark call-to-action (CTA) som uppmanar tittarna att registrera sig pa RankLeague via influencerns referral-lank eller referral-kod.",
        "CTA:n ska vara verbal (sags hogt i videon) och visuell (visas pa skarmen).",
        "Influencern ska namnge RankLeague och kort forklara tjanstens varde.",
        f"Referral-koden <b>{data.get('referral_kod', '')}</b> eller referral-lanken ska vara tydligt synlig i videon och i videobeskrivningen.",
        "Innehallet far inte strida mot svensk lag eller RankLeagues varderingar.",
    ]
    for k in krav:
        story.append(Paragraph(f"\u2022  {k}", styles['BulletItem']))

    # === RAPPORTERING ===
    story.append(Paragraph("5. Rapportering", styles['SectionHeader']))
    story.append(Paragraph(
        "Influencern ska inom <b>7 dagar</b> efter varje publicerad video dela foljande statistik med Uppdragsgivaren:",
        styles['BodyText2']
    ))
    rapport_items = [
        "Antal visningar",
        "Antal klick pa referral-lank (om tillgangligt)",
        "Skarmbild pa videostatistik fran plattformens analys",
    ]
    for item in rapport_items:
        story.append(Paragraph(f"\u2022  {item}", styles['BulletItem']))

    # === AVTALSTID ===
    story.append(Paragraph("6. Avtalstid", styles['SectionHeader']))

    avtalstid = data.get('avtalstid_dagar', 30)
    datum_text = data.get('datum', datetime.now().strftime('%Y-%m-%d'))

    story.append(Paragraph(
        f"Avtalet galler i <b>{avtalstid} dagar</b> fran och med signeringsdatum. "
        f"Om inget annat overenskommits upph\u00f6r avtalet automatiskt efter avtalstiden.",
        styles['BodyText2']
    ))

    if data.get('expires_at'):
        story.append(Paragraph(
            f"Utgangsdatum: <b>{data['expires_at'][:10]}</b>",
            styles['BodyText2']
        ))

    # === BETALNINGSVILLKOR ===
    story.append(Paragraph("7. Betalningsvillkor", styles['SectionHeader']))
    story.append(Paragraph(
        "Utbetalning sker inom 30 dagar efter att influencern levererat statistik for respektive video. "
        "Provisionsersattning (signups) utbetalas manadsvis baserat pa data fran RankLeagues system.",
        styles['BodyText2']
    ))

    # === SIGNERING ===
    story.append(Spacer(1, 24))
    story.append(HRFlowable(width="100%", thickness=1, color=BORDER_GRAY, spaceBefore=8, spaceAfter=16))

    story.append(Paragraph("Signering", styles['SectionHeader']))
    story.append(Paragraph(
        f"Genom att signera detta avtal godkanner bada parter ovanstaende villkor.",
        styles['BodyText2']
    ))
    story.append(Paragraph(
        f"Datum: {datum_text}",
        styles['BodyText2']
    ))

    story.append(Spacer(1, 20))

    # Signerings-rader
    sig_data = [
        [
            f"Uppdragsgivare:\n{data.get('foretag_namn', '')}\n{data.get('kontaktperson', '')}",
            f"Influencer:\n{data.get('influencer_namn', '')}\n@{data.get('kanalnamn', '')}",
        ],
        ['', ''],
        ['____________________________', '____________________________'],
        ['Underskrift', 'Underskrift'],
    ]

    sig_table = Table(sig_data, colWidths=[220, 220])
    sig_table.setStyle(TableStyle([
        ('FONTNAME', (0, 0), (-1, -1), 'Helvetica'),
        ('FONTSIZE', (0, 0), (-1, 1), 10),
        ('FONTSIZE', (0, 2), (-1, -1), 9),
        ('TEXTCOLOR', (0, 0), (-1, -1), DARK_TEXT),
        ('TEXTCOLOR', (0, 3), (-1, 3), GRAY_TEXT),
        ('ALIGN', (0, 0), (-1, -1), 'LEFT'),
        ('TOPPADDING', (0, 0), (-1, -1), 4),
        ('BOTTOMPADDING', (0, 0), (-1, -1), 4),
    ]))
    story.append(sig_table)

    # Footer
    story.append(Spacer(1, 24))
    story.append(HRFlowable(width="100%", thickness=0.5, color=BORDER_GRAY, spaceBefore=8, spaceAfter=8))
    story.append(Paragraph(
        f"Genererat via SparkCollab.se \u2022 {datum_text}",
        ParagraphStyle('Footer', parent=styles['Normal'],
                       fontSize=8, textColor=GRAY_TEXT, alignment=TA_CENTER)
    ))

    # Bygg PDF
    doc.build(story)
    return output_path


if __name__ == '__main__':
    data = json.loads(sys.stdin.read())
    path = generate_pdf(data)
    print(json.dumps({"status": "ok", "path": path}))
